Sources/NIOHTTP2/ConnectionStateMachine/ConnectionStateMachine.swift

```diff
@@ -0,0 +1,1283 @@
+            self.state = .bothQuiescing(state)
+            return result
+
+        case .quiescingPrefaceReceived(var state):
+            let result = state.receivePushPromise(originalStreamID: originalStreamID, childStreamID: childStreamID, headers: headers)
+            self.state = .quiescingPrefaceReceived(state)
+            return result
+
+        case .idle, .prefaceSent, .quiescingPrefaceSent:
+            // We're waiting for the remote preface.
+            return .connectionError(underlyingError: NIOHTTP2Errors.MissingPreface(), type: .protocolError)
+
+        case .closed:
+            return .connectionError(underlyingError: NIOHTTP2Errors.IOOnClosedConnection(), type: .protocolError)
+        }
+    }
+
+    mutating func sendPushPromise(originalStreamID: HTTP2StreamID, childStreamID: HTTP2StreamID, headers: HTTPHeaders) -> StateMachineResult {
+        switch self.state {
+        case .prefaceSent(var state):
+            let result = state.sendPushPromise(originalStreamID: originalStreamID, childStreamID: childStreamID, headers: headers)
+            self.state = .prefaceSent(state)
+            return result
+
+        case .active(var state):
+            let result = state.sendPushPromise(originalStreamID: originalStreamID, childStreamID: childStreamID, headers: headers)
+            self.state = .active(state)
+            return result
+
+        case .locallyQuiesced(var state):
+            let result = state.sendPushPromise(originalStreamID: originalStreamID, childStreamID: childStreamID, headers: headers)
+            self.state = .locallyQuiesced(state)
+            return result
+
+        case .remotelyQuiesced, .bothQuiescing:
+            // We have been quiesced, and may not create new streams.
+            return .connectionError(underlyingError: NIOHTTP2Errors.CreatedStreamAfterGoaway(), type: .protocolError)
+
+        case .quiescingPrefaceSent(var state):
+            let result = state.sendPushPromise(originalStreamID: originalStreamID, childStreamID: childStreamID, headers: headers)
+            self.state = .quiescingPrefaceSent(state)
+            return result
+
+        case .idle, .prefaceReceived, .quiescingPrefaceReceived:
+            // We're waiting for the local preface.
+            return .connectionError(underlyingError: NIOHTTP2Errors.MissingPreface(), type: .protocolError)
+
+        case .closed:
+            return .connectionError(underlyingError: NIOHTTP2Errors.IOOnClosedConnection(), type: .protocolError)
+        }
+    }
+
+    /// Called when a PING frame has been received from the network.
+    mutating func receivePing() -> StateMachineResult {
+        // Pings are pretty straightforward: they're basically always allowed. This is a bit weird, but I can find no text in
+        // RFC 7540 that says that receiving PINGs with ACK flags set when no PING ACKs are expected is forbidden. This is
+        // very strange, but we allow it.
```

Comment:
Not sure I understand your concern correctly, but: as far as I remember a PING with ACK https://tools.ietf.org/html/rfc7540#section-6.7 can be seen kind of like a "PONG" in WebSockets, where peers may send those at-will, uni-directional, not expecting anything back, though perhaps utilizing this as a form of heartbeat - related section in WS: https://tools.ietf.org/html/rfc6455#section-5.5.3 I assumed the receiving and ignoring of PING+ACK at any time would serve a similar purpose here.